### PR TITLE
Fix queries with offset but without limit

### PR DESCRIPTION
**Pull request by matriv**
## Summary of the changes / Why this improves CrateDB

See commit messages

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Review by mfussenegger — COMMENTED** [review]
What do you think of extending `TopNProjection` to support NO_LIMIT?

Compared to the current solution it would have the advantage that the Planner & ProjectionBuilder wouldn't have to special case it.
The ProjectionToProjectionVisitor would then create a SkipProjector within the `visitTopNProjection`. 
1 place to handle it vs. 3

**Comment by matriv** [comment]
I kinda liked more the current PR's approach because of the naming as well: TopNProjection without limit, sounded a bit weird, but if you think it's better to save us some boilerplate code, I'm also fine, I can just remove the 
`assert limit > TopN.NO_LIMIT : "limit of TopNProjection must not be negative/unlimited";`  from the ctor of `TopNProjection`.

BTW, do you think it deserves an integ test as well?

**Comment by mfussenegger** [comment]
> TopNProjection without limit, sounded a bit weird, but if you think it's better to save us some boilerplate code

It's not the boilerplate I'm worried about, but that several components need to know about the limitations of each other. Poor coupling and error prone.

We could rename `TopNProjection` to `LimitAndOffsetProjection` or something like that.

File: server/src/test/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitorTest.java [hunk 1]
```diff
@@ -136,25 +136,42 @@ public void testSimpleTopNProjection() throws Exception {
...
     @Test
-    public void testTopNProjectionToSortingProjector() throws Exception {
+    public void testNoLimitNorOffsetProjection() throws Exception {
+        LimitAndOffsetProjection projection = new LimitAndOffsetProjection(-1, 10, Collections.singletonList(DataTypes.LONG));
+
+        Projector projector = visitor.create(
+            projection, txnCtx, RamAccounting.NO_ACCOUNTING, memoryManager, UUID.randomUUID());
+        assertThat(projector, instanceOf(SimpleOffsetProjector.class));
+
+        TestingRowConsumer consumer = new TestingRowConsumer();
```
Comment (matriv) [inline server/src/test/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitorTest.java:159]: Maybe we should remove this test, and the similar ones from the class, as they are tested in the respective Projectors?

File: libs/dex/src/main/java/io/crate/data/LimitDistinctBatchIterator.java [hunk 2]
```diff
@@ -27,14 +27,14 @@
...
-public final class TopNDistinctBatchIterator<T> implements BatchIterator<T> {
+public final class LimitDistinctBatchIterator<T> implements BatchIterator<T> {
 
     private final BatchIterator<T> source;
     private final Set<T> items;
     private final int limit;
     private final Function<? super T, ? extends T> memoizeItem;
 
-    public TopNDistinctBatchIterator(BatchIterator<T> source, int limit, Function<? super T, ? extends T> memoizeItem) {
+    public LimitDistinctBatchIterator(BatchIterator<T> source, int limit, Function<? super T, ? extends T> memoizeItem) {
```
Comment (mfussenegger) [inline libs/dex/src/main/java/io/crate/data/LimitDistinctBatchIterator.java:37]: Could leave the name as is. TopN still fits for this.

File: server/src/main/java/io/crate/execution/engine/pipeline/LimitDistinctProjector.java [hunk 3]
```diff
@@ -26,21 +26,21 @@
 import io.crate.data.Projector;
 import io.crate.data.Row;
 import io.crate.data.RowN;
-import io.crate.data.TopNDistinctBatchIterator;
+import io.crate.data.LimitDistinctBatchIterator;
 
-public final class TopNDistinctProjector implements Projector {
+public final class LimitDistinctProjector implements Projector {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/pipeline/LimitDistinctProjector.java:31]: Same here, could keep the name for this, it still fits

File: server/src/main/java/io/crate/execution/engine/pipeline/SimpleLimitAndOffsetProjector.java [hunk 4]
```diff
@@ -27,12 +27,12 @@
 import io.crate.data.Row;
 import io.crate.data.SkippingBatchIterator;
 
-public class SimpleTopNProjector implements Projector {
+public class SimpleLimitAndOffsetProjector implements Projector {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/pipeline/SimpleLimitAndOffsetProjector.java:30]:
```suggestion
public class LimitAndOffsetProjector implements Projector {
```

There is no `ComplexLimitAndOffsetProjector` the Simple prefix is not really adding meaningful information.

File: server/src/main/java/io/crate/execution/engine/pipeline/SimpleOffsetProjector.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    public BatchIterator<Row> apply(BatchIterator<Row> batchIterator) {
+        if (batchIterator == null) {
+            return null;
+        }
+        if (offset > 0) {
+            return new SkippingBatchIterator<>(batchIterator, offset);
+        }
+        return batchIterator;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/pipeline/SimpleOffsetProjector.java:48]:
```suggestion
    return new SkippingBatchIterator<>(batchIterator, offset);
```

Afaik the batchIterator is never null and the constructor already asserts that the offset is > 0

File: server/src/test/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitorTest.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/test/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitorTest.java:159]: If it's already tested elsewhere there is no need to test the same thing here too

**Review by mfussenegger — COMMENTED** [review]
I'd tend to do the rename in a separate PR that's not getting backported.
Some of the optimizer rule names change. Not a big deal, even for a hotfix, but given that we can easily avoid it I would

**Comment by matriv** [comment]
> I'd tend to do the rename in a separate PR that's not getting backported. Some of the optimizer rule names change. Not a big deal, even for a hotfix, but given that we can easily avoid it I would

But since it's a different commit I can skip from backporting without a new PR, but if you prefer a separate PR, no problem as well!

File: server/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java [hunk 6]
```diff
@@ -287,19 +292,25 @@ public Projector visitOrderedTopN(OrderedTopNProjection projection, Context cont
...
 
     @Override
-    public Projector visitTopNProjection(TopNProjection projection, Context context) {
-        assert projection.limit() > TopN.NO_LIMIT : "TopNProjection must have a limit";
-        return new SimpleTopNProjector(projection.limit(), projection.offset());
+    public Projector visitLimitAndOffsetProjection(LimitAndOffsetProjection projection, Context context) {
+        if (projection.limit() != NO_LIMIT) {
+            return new SimpleLimitAndOffsetProjector(projection.limit(), projection.offset());
+        } else if (projection.offset() != NO_OFFSET) {
+            return new SimpleOffsetProjector(projection.offset());
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java:309]: Could also consider to have a `LimitProjector` and use both if there is limit-and-offset

File: server/src/main/java/io/crate/execution/engine/pipeline/SimpleOffsetProjector.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+package io.crate.execution.engine.pipeline;
+
+import io.crate.data.BatchIterator;
+import io.crate.data.Projector;
+import io.crate.data.Row;
+import io.crate.data.SkippingBatchIterator;
+
+public class SimpleOffsetProjector implements Projector {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/pipeline/SimpleOffsetProjector.java:29]:
```suggestion
public class OffsetProjector implements Projector {
```

**Comment by mfussenegger** [comment]
> But since it's a different commit I can skip from backporting without a new PR, but if you prefer a separate PR, no problem as well!

Either way is fine. If you prefer to do the backport manually that's okay too

File: server/src/test/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitorTest.java (hunk 1 above)
Comment (matriv) [reply server/src/test/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitorTest.java:159]: Thx, I will double check it and maybe remove in a separate PR (I mean the existing tests)

File: server/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java (hunk 6 above)
Comment (matriv) [reply server/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java]: with the current infrastructure Is there a nice way to do, and not resort to:
```
@Override
    public Projector visitLimitAndOffsetProjection(LimitAndOffsetProjection projection, Context context) {
        final Projector[] projector = new Projector[1];
        if (projection.limit() != NO_LIMIT) {
            projector[0] = new LimitProjector(projection.limit());
        }
        if (projection.offset() != NO_OFFSET) {
            if (projector[0] == null) {
                projector[0] = new OffsetProjector(projection.offset());
            } else {
                projector[0] = rowBatchIterator -> new OffsetProjector(projection.offset()).apply(projector[0].apply(
                    rowBatchIterator));
            }
        }
        if (projector[0] == null) {
            throw new IllegalArgumentException("LimitAndOffsetProjection should have at least one " +
                                               "of the limit and offset set");
        }
        return projector[0];
    }
```

**Review by mfussenegger — APPROVED** [review]

**Comment by matriv** [comment]
retest this please

**Comment by matriv** [comment]
retest this please

**Comment by matriv** [comment]
retest this please

**Pull request merged**